### DISCUSSION TITLE
JS performance ?

### DISCUSSION THREAD

**Discussion by Quafadas** [pityka/saddle#522] [General]
I've just come across this library, and it looks at first glance, quite promising! 

My experiments are so far, rather naive. 
`scala-cli run https://gist.github.com/Quafadas/bd7ae9e354786d98d766e9f5ff92b3e6`

But one of the big things I've often wanted, is a cross platform, basic (my needs are pretty simple) vector library. I had a look on maven, and it looks the core module, is cross published to scala JS. Which is really cool. A couple of questions: 

- Do you know, if the maintainers have a need for the JS support? Or is it more of a side concern?
- Has JS performance been tested at all?

**Comment by pityka** [comment]
Hi! Thanks for your interest!

- JS support of the core modules is not a side concern. All the semantic test suite pass on JS in the ci. I use it regularly on JS for small data sizes both for visualization and lightweight numeric calculations.
- JS performance lacks behind and JVM performance takes precedence. JS performance is not analyzed. Running the test suite in node vs in the jvm reveals that there are huge differences. 
- 'vector library': if you mean linalg, then the saddle-linalg module is not published for JS due to its dependency on native blas and lapack, and I don't see a way out of that. Otherwise the core module contains Vec and Mat.

From the gist I see that you tried it on data visualizations. Saddle works nicely with nspl (https://github.com/pityka/nspl)  as well.

**Comment by Quafadas** [comment]
Wow - thanks for the fast response. 

> JS support of the core modules is not a side concern. All the semantic test suite pass on JS in the ci. I use it regularly on JS for small data sizes both for visualization and lightweight numeric calculations.

:-). I think my needs, are very basic. I don't think I need the lin alg part - in core is enough for me I think. 

> From the gist I see that you tried it on data visualizations. Saddle works nicely with nspl (https://github.com/pityka/nspl) as well.

Cool lib! My personal interest has been trying to make plotting in scala practical for a while - the data visualisation is something that I want to do myself using Vega. I think I have quite a nice scala wrapper for it now. 

Final question: 

> JS performance lacks behind and JVM performance takes precedence. JS performance is not analyzed. Running the test suite in node vs in the jvm reveals that there are huge differences.

Understood. I had a browse through the code. My reading of the `Vec` implementation, is that many of the operations have quite a fundamental reliance on this `raw` method. `BinOps` and many of the underling methods call it repeatedly in loops.
https://github.com/Quafadas/saddle/blob/4923831f44a09de62d0782c98e92fda2e6cdd66c/saddle-core/src/main/scala/org/saddle/vec/VecDefault.scala#L50

I think it's used almost everywhere. I _think_ that in scalaJS, that can be slow... 
https://github.com/raquo/ew/blob/master/README.md

What I'm wondering, is whether or not it may be possible to have a JS implementation of`VecDefault`, backed by 
https://github.com/raquo/ew/blob/master/src/main/scala/com/raquo/ew/JsArray.scala 

instead of a "normal" `array`. My hypothesis, is that just changing that 'raw' method, to use the native browser index via
https://github.com/raquo/ew/blob/b490767a482d0f383ed81a6dd326cb6b94f89500/src/main/scala/com/raquo/ew/JsArray.scala#L117

could result in some significant performance gains, whilst leaving both the JVM implementation, and everything beyond that point, untouched...

I honestly don't know if I'll get the time to try, but if I got the tests to pass and it did exhibit some improvement, would you consider merging that PR? I think the downsides would be 

- build dependance on an extra library for the JS artefact `ew` 
- complexity - as you'd have a planform dependant implementation of VecDefault.

Also, it may not be possible! Do you think it's worth pursuing?

**Reply by pityka** [reply]
I think it is worth a try. It is likely this could be done and could bring some boost on js. Will probably need some work around Vec.apply.

Build complexity is not a problem if you figure out how to set it up :) 
However I prefer to copy the relevant part of that JsArray.scala in the tree (with proper copyright attributions to the author) rather a dependency.

**Reply by Quafadas** [reply]
Well, I gave this a shot... 
https://github.com/Quafadas/saddle/tree/js_perf_test

This compiles, and passes all but 1 of the tests (I commented it out for now). I believe there is a slight positive performance, on the Vec class. 

On the master branch, VecCheck takes about 635ms. Anecdotally, on this branch, it's about 515ms. So a good speedup. However, it had some terrible other implications. 

Firstly, one test failed - which means this somehow changed the semantics - I commented that test out - please beware (!) - for now). 

Secondly, the Mat tests, took about 7x longer. So I think the overhead of the native array construction, is only a net win, if it's used throughout, in the Array, Mat and Vec classes all together. I conclude my idea, was too naive.

Unless one _stays_ in native array land throughout and puts it in the API itself, then I think the downsides significantly outweigh the benefits ...  (and I think this would be a highly risky, major breaking change, which I don't think either of us have appetite for). 

So it's case closed for me! Thanks for being willing to write back - I had some fun looking through the code, and am familiar enough to  give this a test in my use case, in the hope that the performance is already good enough...
